// @polymathnetwork/polymesh-sdk/polkadot/definitions.js
"use strict";
/* istanbul ignore file */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.asset = exports.staking = exports.protocolFee = exports.pips = exports.identity = exports.compliance = exports.polymesh = void 0;
var definitions_1 = require("./polymesh/definitions");
Object.defineProperty(exports, "polymesh", { enumerable: true, get: function () { return __importDefault(definitions_1).default; } });
var definitions_2 = require("./compliance/definitions");
Object.defineProperty(exports, "compliance", { enumerable: true, get: function () { return __importDefault(definitions_2).default; } });
var definitions_3 = require("./identity/definitions");
Object.defineProperty(exports, "identity", { enumerable: true, get: function () { return __importDefault(definitions_3).default; } });
var definitions_4 = require("./pips/definitions");
Object.defineProperty(exports, "pips", { enumerable: true, get: function () { return __importDefault(definitions_4).default; } });
var definitions_5 = require("./protocolFee/definitions");
Object.defineProperty(exports, "protocolFee", { enumerable: true, get: function () { return __importDefault(definitions_5).default; } });
var definitions_6 = require("./staking/definitions");
Object.defineProperty(exports, "staking", { enumerable: true, get: function () { return __importDefault(definitions_6).default; } });
var definitions_7 = require("./asset/definitions");
Object.defineProperty(exports, "asset", { enumerable: true, get: function () { return __importDefault(definitions_7).default; } });
