// @polymathnetwork/polymesh-sdk/middleware/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionOrderFields = exports.StakingEventIdEnum = exports.SnapshotEnum = exports.SettlementResultEnum = exports.SettlementDirectionEnum = exports.ProposalVotesOrderFields = exports.ProposalStateEnum = exports.ProposalOrderFields = exports.Order = exports.ModuleIdEnum = exports.ItnRewardActionType = exports.EventIdEnum = exports.ClaimTypeEnum = exports.ClaimScopeTypeEnum = exports.CallIdEnum = exports.CacheControlScope = exports.AuthTypeEnum = exports.AuthStatusEnum = void 0;
var AuthStatusEnum;
(function (AuthStatusEnum) {
    AuthStatusEnum["Pending"] = "Pending";
    AuthStatusEnum["Consumed"] = "Consumed";
    AuthStatusEnum["Rejected"] = "Rejected";
    AuthStatusEnum["Revoked"] = "Revoked";
    AuthStatusEnum["Expired"] = "Expired";
})(AuthStatusEnum = exports.AuthStatusEnum || (exports.AuthStatusEnum = {}));
var AuthTypeEnum;
(function (AuthTypeEnum) {
    AuthTypeEnum["AttestPrimaryKeyRotation"] = "AttestPrimaryKeyRotation";
    AuthTypeEnum["RotatePrimaryKey"] = "RotatePrimaryKey";
    AuthTypeEnum["RotatePrimaryKeyToSecondary"] = "RotatePrimaryKeyToSecondary";
    AuthTypeEnum["TransferTicker"] = "TransferTicker";
    AuthTypeEnum["AddMultiSigSigner"] = "AddMultiSigSigner";
    AuthTypeEnum["TransferAssetOwnership"] = "TransferAssetOwnership";
    AuthTypeEnum["JoinIdentity"] = "JoinIdentity";
    AuthTypeEnum["PortfolioCustody"] = "PortfolioCustody";
    AuthTypeEnum["BecomeAgent"] = "BecomeAgent";
    AuthTypeEnum["AddRelayerPayingKey"] = "AddRelayerPayingKey";
    AuthTypeEnum["TransferPrimaryIssuanceAgent"] = "TransferPrimaryIssuanceAgent";
    AuthTypeEnum["Custom"] = "Custom";
    AuthTypeEnum["NoData"] = "NoData";
})(AuthTypeEnum = exports.AuthTypeEnum || (exports.AuthTypeEnum = {}));
var CacheControlScope;
(function (CacheControlScope) {
    CacheControlScope["Public"] = "PUBLIC";
    CacheControlScope["Private"] = "PRIVATE";
})(CacheControlScope = exports.CacheControlScope || (exports.CacheControlScope = {}));
var CallIdEnum;
(function (CallIdEnum) {
    CallIdEnum["FillBlock"] = "fill_block";
    CallIdEnum["Remark"] = "remark";
    CallIdEnum["SetHeapPages"] = "set_heap_pages";
    CallIdEnum["SetCode"] = "set_code";
    CallIdEnum["SetCodeWithoutChecks"] = "set_code_without_checks";
    CallIdEnum["SetChangesTrieConfig"] = "set_changes_trie_config";
    CallIdEnum["SetStorage"] = "set_storage";
    CallIdEnum["KillStorage"] = "kill_storage";
    CallIdEnum["KillPrefix"] = "kill_prefix";
    CallIdEnum["Suicide"] = "suicide";
    CallIdEnum["ReportEquivocation"] = "report_equivocation";
    CallIdEnum["ReportEquivocationUnsigned"] = "report_equivocation_unsigned";
    CallIdEnum["Set"] = "set";
    CallIdEnum["Claim"] = "claim";
    CallIdEnum["Transfer"] = "transfer";
    CallIdEnum["Free"] = "free";
    CallIdEnum["ForceTransfer"] = "force_transfer";
    CallIdEnum["Freeze"] = "freeze";
    CallIdEnum["TransferWithMemo"] = "transfer_with_memo";
    CallIdEnum["DepositBlockRewardReserveBalance"] = "deposit_block_reward_reserve_balance";
    CallIdEnum["SetBalance"] = "set_balance";
    CallIdEnum["BurnAccountBalance"] = "burn_account_balance";
    CallIdEnum["SetUncles"] = "set_uncles";
    CallIdEnum["Bond"] = "bond";
    CallIdEnum["BondExtra"] = "bond_extra";
    CallIdEnum["Unbond"] = "unbond";
    CallIdEnum["WithdrawUnbonded"] = "withdraw_unbonded";
    CallIdEnum["Validate"] = "validate";
    CallIdEnum["Nominate"] = "nominate";
    CallIdEnum["Chill"] = "chill";
    CallIdEnum["SetPayee"] = "set_payee";
    CallIdEnum["SetController"] = "set_controller";
    CallIdEnum["SetValidatorCount"] = "set_validator_count";
    CallIdEnum["IncreaseValidatorCount"] = "increase_validator_count";
    CallIdEnum["ScaleValidatorCount"] = "scale_validator_count";
    CallIdEnum["AddPermissionedValidator"] = "add_permissioned_validator";
    CallIdEnum["RemovePermissionedValidator"] = "remove_permissioned_validator";
    CallIdEnum["ValidateCddExpiryNominators"] = "validate_cdd_expiry_nominators";
    CallIdEnum["SetCommissionCap"] = "set_commission_cap";
    CallIdEnum["EnableIndividualCommissions"] = "enable_individual_commissions";
    CallIdEnum["SetGlobalCommission"] = "set_global_commission";
    CallIdEnum["SetMinBondThreshold"] = "set_min_bond_threshold";
    CallIdEnum["ForceNoEras"] = "force_no_eras";
    CallIdEnum["ForceNewEra"] = "force_new_era";
    CallIdEnum["SetInvulnerables"] = "set_invulnerables";
    CallIdEnum["ForceUnstake"] = "force_unstake";
    CallIdEnum["ForceNewEraAlways"] = "force_new_era_always";
    CallIdEnum["CancelDeferredSlash"] = "cancel_deferred_slash";
    CallIdEnum["PayoutStakers"] = "payout_stakers";
    CallIdEnum["Rebond"] = "rebond";
    CallIdEnum["SetHistoryDepth"] = "set_history_depth";
    CallIdEnum["ReapStash"] = "reap_stash";
    CallIdEnum["SubmitElectionSolution"] = "submit_election_solution";
    CallIdEnum["SubmitElectionSolutionUnsigned"] = "submit_election_solution_unsigned";
    CallIdEnum["PayoutStakersBySystem"] = "payout_stakers_by_system";
    CallIdEnum["ChangeSlashingAllowedFor"] = "change_slashing_allowed_for";
    CallIdEnum["SetKeys"] = "set_keys";
    CallIdEnum["PurgeKeys"] = "purge_keys";
    CallIdEnum["FinalHint"] = "final_hint";
    CallIdEnum["NoteStalled"] = "note_stalled";
    CallIdEnum["Heartbeat"] = "heartbeat";
    CallIdEnum["SetSlashingParams"] = "set_slashing_params";
    CallIdEnum["Sudo"] = "sudo";
    CallIdEnum["SudoUncheckedWeight"] = "sudo_unchecked_weight";
    CallIdEnum["SetKey"] = "set_key";
    CallIdEnum["SudoAs"] = "sudo_as";
    CallIdEnum["CreateMultisig"] = "create_multisig";
    CallIdEnum["CreateOrApproveProposalAsIdentity"] = "create_or_approve_proposal_as_identity";
    CallIdEnum["CreateOrApproveProposalAsKey"] = "create_or_approve_proposal_as_key";
    CallIdEnum["CreateProposalAsIdentity"] = "create_proposal_as_identity";
    CallIdEnum["CreateProposalAsKey"] = "create_proposal_as_key";
    CallIdEnum["ApproveAsIdentity"] = "approve_as_identity";
    CallIdEnum["ApproveAsKey"] = "approve_as_key";
    CallIdEnum["RejectAsIdentity"] = "reject_as_identity";
    CallIdEnum["RejectAsKey"] = "reject_as_key";
    CallIdEnum["AcceptMultisigSignerAsIdentity"] = "accept_multisig_signer_as_identity";
    CallIdEnum["AcceptMultisigSignerAsKey"] = "accept_multisig_signer_as_key";
    CallIdEnum["AddMultisigSigner"] = "add_multisig_signer";
    CallIdEnum["RemoveMultisigSigner"] = "remove_multisig_signer";
    CallIdEnum["AddMultisigSignersViaCreator"] = "add_multisig_signers_via_creator";
    CallIdEnum["RemoveMultisigSignersViaCreator"] = "remove_multisig_signers_via_creator";
    CallIdEnum["ChangeSigsRequired"] = "change_sigs_required";
    CallIdEnum["ChangeAllSignersAndSigsRequired"] = "change_all_signers_and_sigs_required";
    CallIdEnum["MakeMultisigSigner"] = "make_multisig_signer";
    CallIdEnum["MakeMultisigPrimary"] = "make_multisig_primary";
    CallIdEnum["ExecuteScheduledProposal"] = "execute_scheduled_proposal";
    CallIdEnum["UpdateSchedule"] = "update_schedule";
    CallIdEnum["SetPutCodeFlag"] = "set_put_code_flag";
    CallIdEnum["PutCode"] = "put_code";
    CallIdEnum["Call"] = "call";
    CallIdEnum["Instantiate"] = "instantiate";
    CallIdEnum["ClaimSurcharge"] = "claim_surcharge";
    CallIdEnum["FreezeInstantiation"] = "freeze_instantiation";
    CallIdEnum["UnfreezeInstantiation"] = "unfreeze_instantiation";
    CallIdEnum["TransferTemplateOwnership"] = "transfer_template_ownership";
    CallIdEnum["ChangeTemplateFees"] = "change_template_fees";
    CallIdEnum["ChangeTemplateMetaUrl"] = "change_template_meta_url";
    CallIdEnum["Disbursement"] = "disbursement";
    CallIdEnum["Reimbursement"] = "reimbursement";
    CallIdEnum["SetVoteThreshold"] = "set_vote_threshold";
    CallIdEnum["Close"] = "close";
    CallIdEnum["SetReleaseCoordinator"] = "set_release_coordinator";
    CallIdEnum["VoteEnactReferendum"] = "vote_enact_referendum";
    CallIdEnum["VoteRejectReferendum"] = "vote_reject_referendum";
    CallIdEnum["SetExpiresAfter"] = "set_expires_after";
    CallIdEnum["VoteOrPropose"] = "vote_or_propose";
    CallIdEnum["Vote"] = "vote";
    CallIdEnum["SetActiveMembersLimit"] = "set_active_members_limit";
    CallIdEnum["DisableMember"] = "disable_member";
    CallIdEnum["AddMember"] = "add_member";
    CallIdEnum["RemoveMember"] = "remove_member";
    CallIdEnum["SwapMember"] = "swap_member";
    CallIdEnum["ResetMembers"] = "reset_members";
    CallIdEnum["AbdicateMembership"] = "abdicate_membership";
    CallIdEnum["SetPruneHistoricalPips"] = "set_prune_historical_pips";
    CallIdEnum["SetMinProposalDeposit"] = "set_min_proposal_deposit";
    CallIdEnum["SetQuorumThreshold"] = "set_quorum_threshold";
    CallIdEnum["SetProposalDuration"] = "set_proposal_duration";
    CallIdEnum["SetProposalCoolOffPeriod"] = "set_proposal_cool_off_period";
    CallIdEnum["SetDefaultEnactmentPeriod"] = "set_default_enactment_period";
    CallIdEnum["SetPendingPipExpiry"] = "set_pending_pip_expiry";
    CallIdEnum["SetMaxPipSkipCount"] = "set_max_pip_skip_count";
    CallIdEnum["SetActivePipLimit"] = "set_active_pip_limit";
    CallIdEnum["Propose"] = "propose";
    CallIdEnum["AmendProposal"] = "amend_proposal";
    CallIdEnum["CancelProposal"] = "cancel_proposal";
    CallIdEnum["ApproveCommitteeProposal"] = "approve_committee_proposal";
    CallIdEnum["RejectProposal"] = "reject_proposal";
    CallIdEnum["BondAdditionalDeposit"] = "bond_additional_deposit";
    CallIdEnum["UnbondDeposit"] = "unbond_deposit";
    CallIdEnum["KillProposal"] = "kill_proposal";
    CallIdEnum["PruneProposal"] = "prune_proposal";
    CallIdEnum["RescheduleExecution"] = "reschedule_execution";
    CallIdEnum["ClearSnapshot"] = "clear_snapshot";
    CallIdEnum["Snapshot"] = "snapshot";
    CallIdEnum["EnactSnapshotResults"] = "enact_snapshot_results";
    CallIdEnum["FastTrackProposal"] = "fast_track_proposal";
    CallIdEnum["EmergencyReferendum"] = "emergency_referendum";
    CallIdEnum["EnactReferendum"] = "enact_referendum";
    CallIdEnum["RejectReferendum"] = "reject_referendum";
    CallIdEnum["OverrideReferendumEnactmentPeriod"] = "override_referendum_enactment_period";
    CallIdEnum["RegisterTicker"] = "register_ticker";
    CallIdEnum["CreateAssetAndMint"] = "create_asset_and_mint";
    CallIdEnum["AcceptTickerTransfer"] = "accept_ticker_transfer";
    CallIdEnum["AcceptPrimaryIssuanceAgentTransfer"] = "accept_primary_issuance_agent_transfer";
    CallIdEnum["AcceptAssetOwnershipTransfer"] = "accept_asset_ownership_transfer";
    CallIdEnum["CreateAsset"] = "create_asset";
    CallIdEnum["Unfreeze"] = "unfreeze";
    CallIdEnum["RenameAsset"] = "rename_asset";
    CallIdEnum["CreateCheckpoint"] = "create_checkpoint";
    CallIdEnum["Issue"] = "issue";
    CallIdEnum["Redeem"] = "redeem";
    CallIdEnum["MakeDivisible"] = "make_divisible";
    CallIdEnum["BatchAddDocument"] = "batch_add_document";
    CallIdEnum["BatchRemoveDocument"] = "batch_remove_document";
    CallIdEnum["AddDocuments"] = "add_documents";
    CallIdEnum["RemoveDocuments"] = "remove_documents";
    CallIdEnum["SetFundingRound"] = "set_funding_round";
    CallIdEnum["UpdateIdentifiers"] = "update_identifiers";
    CallIdEnum["AddExtension"] = "add_extension";
    CallIdEnum["ArchiveExtension"] = "archive_extension";
    CallIdEnum["UnarchiveExtension"] = "unarchive_extension";
    CallIdEnum["RemovePrimaryIssuanceAgent"] = "remove_primary_issuance_agent";
    CallIdEnum["RemoveSmartExtension"] = "remove_smart_extension";
    CallIdEnum["ClaimClassicTicker"] = "claim_classic_ticker";
    CallIdEnum["ReserveClassicTicker"] = "reserve_classic_ticker";
    CallIdEnum["RegisterCustomAssetType"] = "register_custom_asset_type";
    CallIdEnum["New"] = "new";
    CallIdEnum["Cancel"] = "cancel";
    CallIdEnum["ClaimUnclaimed"] = "claim_unclaimed";
    CallIdEnum["CddRegisterDid"] = "cdd_register_did";
    CallIdEnum["InvalidateCddClaims"] = "invalidate_cdd_claims";
    CallIdEnum["RemoveSecondaryKeys"] = "remove_secondary_keys";
    CallIdEnum["SetPrimaryKey"] = "set_primary_key";
    CallIdEnum["AcceptPrimaryKey"] = "accept_primary_key";
    CallIdEnum["RotatePrimaryKeyToSecondary"] = "rotate_primary_key_to_secondary";
    CallIdEnum["ChangeCddRequirementForMkRotation"] = "change_cdd_requirement_for_mk_rotation";
    CallIdEnum["JoinIdentityAsKey"] = "join_identity_as_key";
    CallIdEnum["LeaveIdentityAsKey"] = "leave_identity_as_key";
    CallIdEnum["AddClaim"] = "add_claim";
    CallIdEnum["BatchAddClaim"] = "batch_add_claim";
    CallIdEnum["RevokeClaim"] = "revoke_claim";
    CallIdEnum["RevokeClaimByIndex"] = "revoke_claim_by_index";
    CallIdEnum["BatchRevokeClaim"] = "batch_revoke_claim";
    CallIdEnum["SetPermissionToSigner"] = "set_permission_to_signer";
    CallIdEnum["LegacySetPermissionToSigner"] = "legacy_set_permission_to_signer";
    CallIdEnum["FreezeSecondaryKeys"] = "freeze_secondary_keys";
    CallIdEnum["UnfreezeSecondaryKeys"] = "unfreeze_secondary_keys";
    CallIdEnum["AddAuthorization"] = "add_authorization";
    CallIdEnum["BatchAddAuthorization"] = "batch_add_authorization";
    CallIdEnum["RemoveAuthorization"] = "remove_authorization";
    CallIdEnum["BatchRemoveAuthorization"] = "batch_remove_authorization";
    CallIdEnum["AcceptAuthorization"] = "accept_authorization";
    CallIdEnum["BatchAcceptAuthorization"] = "batch_accept_authorization";
    CallIdEnum["BatchAddSecondaryKeyWithAuthorization"] = "batch_add_secondary_key_with_authorization";
    CallIdEnum["AddSecondaryKeysWithAuthorization"] = "add_secondary_keys_with_authorization";
    CallIdEnum["AddInvestorUniquenessClaim"] = "add_investor_uniqueness_claim";
    CallIdEnum["GcAddCddClaim"] = "gc_add_cdd_claim";
    CallIdEnum["GcRevokeCddClaim"] = "gc_revoke_cdd_claim";
    CallIdEnum["AddInvestorUniquenessClaimV2"] = "add_investor_uniqueness_claim_v2";
    CallIdEnum["JoinIdentityAsIdentity"] = "join_identity_as_identity";
    CallIdEnum["LeaveIdentityAsIdentity"] = "leave_identity_as_identity";
    CallIdEnum["RevokeOffchainAuthorization"] = "revoke_offchain_authorization";
    CallIdEnum["ForwardedCall"] = "forwarded_call";
    CallIdEnum["ChangeController"] = "change_controller";
    CallIdEnum["ChangeAdmin"] = "change_admin";
    CallIdEnum["ChangeTimelock"] = "change_timelock";
    CallIdEnum["ChangeBridgeLimit"] = "change_bridge_limit";
    CallIdEnum["ChangeBridgeExempted"] = "change_bridge_exempted";
    CallIdEnum["ForceHandleBridgeTx"] = "force_handle_bridge_tx";
    CallIdEnum["BatchForceHandleBridgeTx"] = "batch_force_handle_bridge_tx";
    CallIdEnum["ProposeBridgeTx"] = "propose_bridge_tx";
    CallIdEnum["BatchProposeBridgeTx"] = "batch_propose_bridge_tx";
    CallIdEnum["HandleBridgeTx"] = "handle_bridge_tx";
    CallIdEnum["BatchHandleBridgeTx"] = "batch_handle_bridge_tx";
    CallIdEnum["BatchFreezeTx"] = "batch_freeze_tx";
    CallIdEnum["BatchUnfreezeTx"] = "batch_unfreeze_tx";
    CallIdEnum["FreezeTxs"] = "freeze_txs";
    CallIdEnum["UnfreezeTxs"] = "unfreeze_txs";
    CallIdEnum["HandleScheduledBridgeTx"] = "handle_scheduled_bridge_tx";
    CallIdEnum["AddFreezeAdmin"] = "add_freeze_admin";
    CallIdEnum["RemoveFreezeAdmin"] = "remove_freeze_admin";
    CallIdEnum["RemoveTxs"] = "remove_txs";
    CallIdEnum["AddComplianceRequirement"] = "add_compliance_requirement";
    CallIdEnum["RemoveComplianceRequirement"] = "remove_compliance_requirement";
    CallIdEnum["ReplaceAssetCompliance"] = "replace_asset_compliance";
    CallIdEnum["ResetAssetCompliance"] = "reset_asset_compliance";
    CallIdEnum["PauseAssetCompliance"] = "pause_asset_compliance";
    CallIdEnum["ResumeAssetCompliance"] = "resume_asset_compliance";
    CallIdEnum["AddDefaultTrustedClaimIssuer"] = "add_default_trusted_claim_issuer";
    CallIdEnum["RemoveDefaultTrustedClaimIssuer"] = "remove_default_trusted_claim_issuer";
    CallIdEnum["BatchAddDefaultTrustedClaimIssuer"] = "batch_add_default_trusted_claim_issuer";
    CallIdEnum["BatchRemoveDefaultTrustedClaimIssuer"] = "batch_remove_default_trusted_claim_issuer";
    CallIdEnum["ChangeComplianceRequirement"] = "change_compliance_requirement";
    CallIdEnum["BatchChangeComplianceRequirement"] = "batch_change_compliance_requirement";
    CallIdEnum["AddBallot"] = "add_ballot";
    CallIdEnum["CancelBallot"] = "cancel_ballot";
    CallIdEnum["LaunchSto"] = "launch_sto";
    CallIdEnum["BuyTokens"] = "buy_tokens";
    CallIdEnum["PauseSto"] = "pause_sto";
    CallIdEnum["UnpauseSto"] = "unpause_sto";
    CallIdEnum["ModifyExemptionList"] = "modify_exemption_list";
    CallIdEnum["CreateVenue"] = "create_venue";
    CallIdEnum["UpdateVenue"] = "update_venue";
    CallIdEnum["AddInstruction"] = "add_instruction";
    CallIdEnum["AddAndAuthorizeInstruction"] = "add_and_authorize_instruction";
    CallIdEnum["AuthorizeInstruction"] = "authorize_instruction";
    CallIdEnum["UnauthorizeInstruction"] = "unauthorize_instruction";
    CallIdEnum["AddAndAffirmInstruction"] = "add_and_affirm_instruction";
    CallIdEnum["AffirmInstruction"] = "affirm_instruction";
    CallIdEnum["WithdrawAffirmation"] = "withdraw_affirmation";
    CallIdEnum["RejectInstruction"] = "reject_instruction";
    CallIdEnum["RescheduleInstruction"] = "reschedule_instruction";
    CallIdEnum["AuthorizeWithReceipts"] = "authorize_with_receipts";
    CallIdEnum["AffirmWithReceipts"] = "affirm_with_receipts";
    CallIdEnum["ClaimReceipt"] = "claim_receipt";
    CallIdEnum["UnclaimReceipt"] = "unclaim_receipt";
    CallIdEnum["SetVenueFiltering"] = "set_venue_filtering";
    CallIdEnum["AllowVenues"] = "allow_venues";
    CallIdEnum["DisallowVenues"] = "disallow_venues";
    CallIdEnum["ExecuteScheduledInstruction"] = "execute_scheduled_instruction";
    CallIdEnum["ChangeReceiptValidity"] = "change_receipt_validity";
    CallIdEnum["UpdateVenueType"] = "update_venue_type";
    CallIdEnum["UpdateVenueDetails"] = "update_venue_details";
    CallIdEnum["CreateFundraiser"] = "create_fundraiser";
    CallIdEnum["Invest"] = "invest";
    CallIdEnum["FreezeFundraiser"] = "freeze_fundraiser";
    CallIdEnum["UnfreezeFundraiser"] = "unfreeze_fundraiser";
    CallIdEnum["ModifyFundraiserWindow"] = "modify_fundraiser_window";
    CallIdEnum["Stop"] = "stop";
    CallIdEnum["AddTransferManager"] = "add_transfer_manager";
    CallIdEnum["RemoveTransferManager"] = "remove_transfer_manager";
    CallIdEnum["AddExemptedEntities"] = "add_exempted_entities";
    CallIdEnum["RemoveExemptedEntities"] = "remove_exempted_entities";
    CallIdEnum["ChangeCoefficient"] = "change_coefficient";
    CallIdEnum["ChangeBaseFee"] = "change_base_fee";
    CallIdEnum["Batch"] = "batch";
    CallIdEnum["BatchAtomic"] = "batch_atomic";
    CallIdEnum["BatchOptimistic"] = "batch_optimistic";
    CallIdEnum["RelayTx"] = "relay_tx";
    CallIdEnum["CreatePortfolio"] = "create_portfolio";
    CallIdEnum["DeletePortfolio"] = "delete_portfolio";
    CallIdEnum["MovePortfolioFunds"] = "move_portfolio_funds";
    CallIdEnum["RenamePortfolio"] = "rename_portfolio";
    CallIdEnum["QuitPortfolioCustody"] = "quit_portfolio_custody";
    CallIdEnum["AcceptPortfolioCustody"] = "accept_portfolio_custody";
    CallIdEnum["AddRangeProof"] = "add_range_proof";
    CallIdEnum["AddVerifyRangeProof"] = "add_verify_range_proof";
    CallIdEnum["Schedule"] = "schedule";
    CallIdEnum["ScheduleNamed"] = "schedule_named";
    CallIdEnum["CancelNamed"] = "cancel_named";
    CallIdEnum["ScheduleAfter"] = "schedule_after";
    CallIdEnum["ScheduleNamedAfter"] = "schedule_named_after";
    CallIdEnum["SetMaxDetailsLength"] = "set_max_details_length";
    CallIdEnum["ResetCaa"] = "reset_caa";
    CallIdEnum["SetDefaultTargets"] = "set_default_targets";
    CallIdEnum["SetDefaultWithholdingTax"] = "set_default_withholding_tax";
    CallIdEnum["SetDidWithholdingTax"] = "set_did_withholding_tax";
    CallIdEnum["InitiateCorporateAction"] = "initiate_corporate_action";
    CallIdEnum["LinkCaDoc"] = "link_ca_doc";
    CallIdEnum["RemoveCa"] = "remove_ca";
    CallIdEnum["ChangeRecordDate"] = "change_record_date";
    CallIdEnum["AttachBallot"] = "attach_ballot";
    CallIdEnum["ChangeEnd"] = "change_end";
    CallIdEnum["ChangeMeta"] = "change_meta";
    CallIdEnum["ChangeRcv"] = "change_rcv";
    CallIdEnum["RemoveBallot"] = "remove_ballot";
    CallIdEnum["Distribute"] = "distribute";
    CallIdEnum["PushBenefit"] = "push_benefit";
    CallIdEnum["Reclaim"] = "reclaim";
    CallIdEnum["RemoveDistribution"] = "remove_distribution";
    CallIdEnum["CreateGroup"] = "create_group";
    CallIdEnum["SetGroupPermissions"] = "set_group_permissions";
    CallIdEnum["RemoveAgent"] = "remove_agent";
    CallIdEnum["Abdicate"] = "abdicate";
    CallIdEnum["ChangeGroup"] = "change_group";
    CallIdEnum["AcceptBecomeAgent"] = "accept_become_agent";
    CallIdEnum["SetSchedulesMaxComplexity"] = "set_schedules_max_complexity";
    CallIdEnum["CreateSchedule"] = "create_schedule";
    CallIdEnum["RemoveSchedule"] = "remove_schedule";
    CallIdEnum["RegisterDid"] = "register_did";
    CallIdEnum["MockCddRegisterDid"] = "mock_cdd_register_did";
    CallIdEnum["GetMyDid"] = "get_my_did";
    CallIdEnum["GetCddOf"] = "get_cdd_of";
    CallIdEnum["SetPayingKey"] = "set_paying_key";
    CallIdEnum["AcceptPayingKey"] = "accept_paying_key";
    CallIdEnum["RemovePayingKey"] = "remove_paying_key";
    CallIdEnum["UpdatePolyxLimit"] = "update_polyx_limit";
    CallIdEnum["IncreasePolyxLimit"] = "increase_polyx_limit";
    CallIdEnum["DecreasePolyxLimit"] = "decrease_polyx_limit";
    CallIdEnum["ClaimItnReward"] = "claim_itn_reward";
    CallIdEnum["SetItnRewardStatus"] = "set_itn_reward_status";
    CallIdEnum["ControllerTransfer"] = "controller_transfer";
    CallIdEnum["Approve"] = "approve";
    CallIdEnum["TransferFrom"] = "transfer_from";
    CallIdEnum["BatchIssue"] = "batch_issue";
    CallIdEnum["RedeemFrom"] = "redeem_from";
    CallIdEnum["ControllerRedeem"] = "controller_redeem";
    CallIdEnum["TransferWithData"] = "transfer_with_data";
    CallIdEnum["TransferFromWithData"] = "transfer_from_with_data";
    CallIdEnum["IsIssuable"] = "is_issuable";
    CallIdEnum["IncreaseCustodyAllowance"] = "increase_custody_allowance";
    CallIdEnum["IncreaseCustodyAllowanceOf"] = "increase_custody_allowance_of";
    CallIdEnum["TransferByCustodian"] = "transfer_by_custodian";
    CallIdEnum["SetTreasuryDid"] = "set_treasury_did";
    CallIdEnum["RemoveSigningKeys"] = "remove_signing_keys";
    CallIdEnum["SetMasterKey"] = "set_master_key";
    CallIdEnum["AcceptMasterKey"] = "accept_master_key";
    CallIdEnum["FreezeSigningKeys"] = "freeze_signing_keys";
    CallIdEnum["UnfreezeSigningKeys"] = "unfreeze_signing_keys";
    CallIdEnum["BatchAddSigningKeyWithAuthorization"] = "batch_add_signing_key_with_authorization";
    CallIdEnum["AddActiveRule"] = "add_active_rule";
    CallIdEnum["RemoveActiveRule"] = "remove_active_rule";
    CallIdEnum["ReplaceAssetRules"] = "replace_asset_rules";
    CallIdEnum["ResetActiveRules"] = "reset_active_rules";
    CallIdEnum["PauseAssetRules"] = "pause_asset_rules";
    CallIdEnum["ResumeAssetRules"] = "resume_asset_rules";
    CallIdEnum["ChangeAssetRule"] = "change_asset_rule";
    CallIdEnum["BatchChangeAssetRule"] = "batch_change_asset_rule";
})(CallIdEnum = exports.CallIdEnum || (exports.CallIdEnum = {}));
var ClaimScopeTypeEnum;
(function (ClaimScopeTypeEnum) {
    ClaimScopeTypeEnum["Identity"] = "Identity";
    ClaimScopeTypeEnum["Ticker"] = "Ticker";
    ClaimScopeTypeEnum["Custom"] = "Custom";
})(ClaimScopeTypeEnum = exports.ClaimScopeTypeEnum || (exports.ClaimScopeTypeEnum = {}));
var ClaimTypeEnum;
(function (ClaimTypeEnum) {
    ClaimTypeEnum["Accredited"] = "Accredited";
    ClaimTypeEnum["Affiliate"] = "Affiliate";
    ClaimTypeEnum["BuyLockup"] = "BuyLockup";
    ClaimTypeEnum["SellLockup"] = "SellLockup";
    ClaimTypeEnum["CustomerDueDiligence"] = "CustomerDueDiligence";
    ClaimTypeEnum["KnowYourCustomer"] = "KnowYourCustomer";
    ClaimTypeEnum["Jurisdiction"] = "Jurisdiction";
    ClaimTypeEnum["Exempted"] = "Exempted";
    ClaimTypeEnum["Blocked"] = "Blocked";
    ClaimTypeEnum["InvestorUniqueness"] = "InvestorUniqueness";
    ClaimTypeEnum["NoData"] = "NoData";
    ClaimTypeEnum["InvestorUniquenessV2"] = "InvestorUniquenessV2";
})(ClaimTypeEnum = exports.ClaimTypeEnum || (exports.ClaimTypeEnum = {}));
var EventIdEnum;
(function (EventIdEnum) {
    EventIdEnum["ExtrinsicSuccess"] = "ExtrinsicSuccess";
    EventIdEnum["ExtrinsicFailed"] = "ExtrinsicFailed";
    EventIdEnum["CodeUpdated"] = "CodeUpdated";
    EventIdEnum["NewAccount"] = "NewAccount";
    EventIdEnum["KilledAccount"] = "KilledAccount";
    EventIdEnum["IndexAssigned"] = "IndexAssigned";
    EventIdEnum["IndexFreed"] = "IndexFreed";
    EventIdEnum["IndexFrozen"] = "IndexFrozen";
    EventIdEnum["Endowed"] = "Endowed";
    EventIdEnum["Transfer"] = "Transfer";
    EventIdEnum["BalanceSet"] = "BalanceSet";
    EventIdEnum["AccountBalanceBurned"] = "AccountBalanceBurned";
    EventIdEnum["Reserved"] = "Reserved";
    EventIdEnum["Unreserved"] = "Unreserved";
    EventIdEnum["ReserveRepatriated"] = "ReserveRepatriated";
    EventIdEnum["EraPayout"] = "EraPayout";
    EventIdEnum["Reward"] = "Reward";
    EventIdEnum["Slash"] = "Slash";
    EventIdEnum["OldSlashingReportDiscarded"] = "OldSlashingReportDiscarded";
    EventIdEnum["StakingElection"] = "StakingElection";
    EventIdEnum["SolutionStored"] = "SolutionStored";
    EventIdEnum["Bonded"] = "Bonded";
    EventIdEnum["Unbonded"] = "Unbonded";
    EventIdEnum["Nominated"] = "Nominated";
    EventIdEnum["Withdrawn"] = "Withdrawn";
    EventIdEnum["PermissionedValidatorAdded"] = "PermissionedValidatorAdded";
    EventIdEnum["PermissionedValidatorRemoved"] = "PermissionedValidatorRemoved";
    EventIdEnum["PermissionedValidatorStatusChanged"] = "PermissionedValidatorStatusChanged";
    EventIdEnum["PermissionedIdentityAdded"] = "PermissionedIdentityAdded";
    EventIdEnum["PermissionedIdentityRemoved"] = "PermissionedIdentityRemoved";
    EventIdEnum["InvalidatedNominators"] = "InvalidatedNominators";
    EventIdEnum["CommissionCapUpdated"] = "CommissionCapUpdated";
    EventIdEnum["IndividualCommissionEnabled"] = "IndividualCommissionEnabled";
    EventIdEnum["GlobalCommissionUpdated"] = "GlobalCommissionUpdated";
    EventIdEnum["MinimumBondThresholdUpdated"] = "MinimumBondThresholdUpdated";
    EventIdEnum["RewardPaymentSchedulingInterrupted"] = "RewardPaymentSchedulingInterrupted";
    EventIdEnum["SlashingAllowedForChanged"] = "SlashingAllowedForChanged";
    EventIdEnum["Offence"] = "Offence";
    EventIdEnum["NewSession"] = "NewSession";
    EventIdEnum["NewAuthorities"] = "NewAuthorities";
    EventIdEnum["Paused"] = "Paused";
    EventIdEnum["Resumed"] = "Resumed";
    EventIdEnum["HeartbeatReceived"] = "HeartbeatReceived";
    EventIdEnum["AllGood"] = "AllGood";
    EventIdEnum["SomeOffline"] = "SomeOffline";
    EventIdEnum["SlashingParamsUpdated"] = "SlashingParamsUpdated";
    EventIdEnum["Sudid"] = "Sudid";
    EventIdEnum["KeyChanged"] = "KeyChanged";
    EventIdEnum["SudoAsDone"] = "SudoAsDone";
    EventIdEnum["MultiSigCreated"] = "MultiSigCreated";
    EventIdEnum["ProposalAdded"] = "ProposalAdded";
    EventIdEnum["ProposalExecuted"] = "ProposalExecuted";
    EventIdEnum["MultiSigSignerAdded"] = "MultiSigSignerAdded";
    EventIdEnum["MultiSigSignerAuthorized"] = "MultiSigSignerAuthorized";
    EventIdEnum["MultiSigSignerRemoved"] = "MultiSigSignerRemoved";
    EventIdEnum["MultiSigSignaturesRequiredChanged"] = "MultiSigSignaturesRequiredChanged";
    EventIdEnum["ProposalApproved"] = "ProposalApproved";
    EventIdEnum["ProposalRejectionVote"] = "ProposalRejectionVote";
    EventIdEnum["ProposalRejected"] = "ProposalRejected";
    EventIdEnum["ProposalExecutionFailed"] = "ProposalExecutionFailed";
    EventIdEnum["Instantiated"] = "Instantiated";
    EventIdEnum["Evicted"] = "Evicted";
    EventIdEnum["Restored"] = "Restored";
    EventIdEnum["CodeStored"] = "CodeStored";
    EventIdEnum["ScheduleUpdated"] = "ScheduleUpdated";
    EventIdEnum["ContractExecution"] = "ContractExecution";
    EventIdEnum["InstantiationFeeChanged"] = "InstantiationFeeChanged";
    EventIdEnum["InstantiationFreezed"] = "InstantiationFreezed";
    EventIdEnum["InstantiationUnFreezed"] = "InstantiationUnFreezed";
    EventIdEnum["TemplateOwnershipTransferred"] = "TemplateOwnershipTransferred";
    EventIdEnum["TemplateUsageFeeChanged"] = "TemplateUsageFeeChanged";
    EventIdEnum["TemplateInstantiationFeeChanged"] = "TemplateInstantiationFeeChanged";
    EventIdEnum["TemplateMetaUrlChanged"] = "TemplateMetaUrlChanged";
    EventIdEnum["PutCodeFlagChanged"] = "PutCodeFlagChanged";
    EventIdEnum["TreasuryDisbursement"] = "TreasuryDisbursement";
    EventIdEnum["TreasuryReimbursement"] = "TreasuryReimbursement";
    EventIdEnum["Proposed"] = "Proposed";
    EventIdEnum["Voted"] = "Voted";
    EventIdEnum["VoteRetracted"] = "VoteRetracted";
    EventIdEnum["FinalVotes"] = "FinalVotes";
    EventIdEnum["Approved"] = "Approved";
    EventIdEnum["Rejected"] = "Rejected";
    EventIdEnum["Executed"] = "Executed";
    EventIdEnum["Closed"] = "Closed";
    EventIdEnum["ReleaseCoordinatorUpdated"] = "ReleaseCoordinatorUpdated";
    EventIdEnum["ExpiresAfterUpdated"] = "ExpiresAfterUpdated";
    EventIdEnum["VoteThresholdUpdated"] = "VoteThresholdUpdated";
    EventIdEnum["VoteEnactReferendum"] = "VoteEnactReferendum";
    EventIdEnum["VoteRejectReferendum"] = "VoteRejectReferendum";
    EventIdEnum["MemberAdded"] = "MemberAdded";
    EventIdEnum["MemberRemoved"] = "MemberRemoved";
    EventIdEnum["MemberRevoked"] = "MemberRevoked";
    EventIdEnum["MembersSwapped"] = "MembersSwapped";
    EventIdEnum["MembersReset"] = "MembersReset";
    EventIdEnum["ActiveLimitChanged"] = "ActiveLimitChanged";
    EventIdEnum["Dummy"] = "Dummy";
    EventIdEnum["HistoricalPipsPruned"] = "HistoricalPipsPruned";
    EventIdEnum["ProposalCreated"] = "ProposalCreated";
    EventIdEnum["ProposalDetailsAmended"] = "ProposalDetailsAmended";
    EventIdEnum["ProposalBondAdjusted"] = "ProposalBondAdjusted";
    EventIdEnum["ProposalStateUpdated"] = "ProposalStateUpdated";
    EventIdEnum["PipClosed"] = "PipClosed";
    EventIdEnum["ExecutionScheduled"] = "ExecutionScheduled";
    EventIdEnum["ReferendumCreated"] = "ReferendumCreated";
    EventIdEnum["ReferendumScheduled"] = "ReferendumScheduled";
    EventIdEnum["ReferendumStateUpdated"] = "ReferendumStateUpdated";
    EventIdEnum["DefaultEnactmentPeriodChanged"] = "DefaultEnactmentPeriodChanged";
    EventIdEnum["MinimumProposalDepositChanged"] = "MinimumProposalDepositChanged";
    EventIdEnum["QuorumThresholdChanged"] = "QuorumThresholdChanged";
    EventIdEnum["ProposalCoolOffPeriodChanged"] = "ProposalCoolOffPeriodChanged";
    EventIdEnum["PendingPipExpiryChanged"] = "PendingPipExpiryChanged";
    EventIdEnum["MaxPipSkipCountChanged"] = "MaxPipSkipCountChanged";
    EventIdEnum["ActivePipLimitChanged"] = "ActivePipLimitChanged";
    EventIdEnum["ProposalDurationChanged"] = "ProposalDurationChanged";
    EventIdEnum["ProposalRefund"] = "ProposalRefund";
    EventIdEnum["SnapshotCleared"] = "SnapshotCleared";
    EventIdEnum["SnapshotTaken"] = "SnapshotTaken";
    EventIdEnum["PipSkipped"] = "PipSkipped";
    EventIdEnum["SnapshotResultsEnacted"] = "SnapshotResultsEnacted";
    EventIdEnum["Approval"] = "Approval";
    EventIdEnum["Issued"] = "Issued";
    EventIdEnum["Redeemed"] = "Redeemed";
    EventIdEnum["ControllerTransfer"] = "ControllerTransfer";
    EventIdEnum["ControllerRedemption"] = "ControllerRedemption";
    EventIdEnum["AssetCreated"] = "AssetCreated";
    EventIdEnum["IdentifiersUpdated"] = "IdentifiersUpdated";
    EventIdEnum["DivisibilityChanged"] = "DivisibilityChanged";
    EventIdEnum["TransferWithData"] = "TransferWithData";
    EventIdEnum["IsIssuable"] = "IsIssuable";
    EventIdEnum["TickerRegistered"] = "TickerRegistered";
    EventIdEnum["TickerTransferred"] = "TickerTransferred";
    EventIdEnum["AssetOwnershipTransferred"] = "AssetOwnershipTransferred";
    EventIdEnum["AssetFrozen"] = "AssetFrozen";
    EventIdEnum["AssetUnfrozen"] = "AssetUnfrozen";
    EventIdEnum["AssetRenamed"] = "AssetRenamed";
    EventIdEnum["FundingRoundSet"] = "FundingRoundSet";
    EventIdEnum["ExtensionAdded"] = "ExtensionAdded";
    EventIdEnum["ExtensionArchived"] = "ExtensionArchived";
    EventIdEnum["ExtensionUnArchive"] = "ExtensionUnArchive";
    EventIdEnum["CheckpointCreated"] = "CheckpointCreated";
    EventIdEnum["PrimaryIssuanceAgentTransferred"] = "PrimaryIssuanceAgentTransferred";
    EventIdEnum["PrimaryIssuanceAgentTransfered"] = "PrimaryIssuanceAgentTransfered";
    EventIdEnum["DocumentAdded"] = "DocumentAdded";
    EventIdEnum["DocumentRemoved"] = "DocumentRemoved";
    EventIdEnum["ExtensionRemoved"] = "ExtensionRemoved";
    EventIdEnum["ClassicTickerClaimed"] = "ClassicTickerClaimed";
    EventIdEnum["CustomAssetTypeExists"] = "CustomAssetTypeExists";
    EventIdEnum["CustomAssetTypeRegistered"] = "CustomAssetTypeRegistered";
    EventIdEnum["InvestorUniquenessClaimNotAllowed"] = "InvestorUniquenessClaimNotAllowed";
    EventIdEnum["DividendCreated"] = "DividendCreated";
    EventIdEnum["DividendCanceled"] = "DividendCanceled";
    EventIdEnum["DividendPaidOutToUser"] = "DividendPaidOutToUser";
    EventIdEnum["DividendRemainingClaimed"] = "DividendRemainingClaimed";
    EventIdEnum["DidCreated"] = "DidCreated";
    EventIdEnum["SecondaryKeysAdded"] = "SecondaryKeysAdded";
    EventIdEnum["SecondaryKeysRemoved"] = "SecondaryKeysRemoved";
    EventIdEnum["SignerLeft"] = "SignerLeft";
    EventIdEnum["SecondaryKeyPermissionsUpdated"] = "SecondaryKeyPermissionsUpdated";
    EventIdEnum["SecondaryPermissionsUpdated"] = "SecondaryPermissionsUpdated";
    EventIdEnum["PrimaryKeyUpdated"] = "PrimaryKeyUpdated";
    EventIdEnum["ClaimAdded"] = "ClaimAdded";
    EventIdEnum["ClaimRevoked"] = "ClaimRevoked";
    EventIdEnum["DidStatus"] = "DidStatus";
    EventIdEnum["CddStatus"] = "CddStatus";
    EventIdEnum["AssetDidRegistered"] = "AssetDidRegistered";
    EventIdEnum["AuthorizationAdded"] = "AuthorizationAdded";
    EventIdEnum["AuthorizationRevoked"] = "AuthorizationRevoked";
    EventIdEnum["AuthorizationRejected"] = "AuthorizationRejected";
    EventIdEnum["AuthorizationConsumed"] = "AuthorizationConsumed";
    EventIdEnum["OffChainAuthorizationRevoked"] = "OffChainAuthorizationRevoked";
    EventIdEnum["CddRequirementForPrimaryKeyUpdated"] = "CddRequirementForPrimaryKeyUpdated";
    EventIdEnum["CddClaimsInvalidated"] = "CddClaimsInvalidated";
    EventIdEnum["SecondaryKeysFrozen"] = "SecondaryKeysFrozen";
    EventIdEnum["SecondaryKeysUnfrozen"] = "SecondaryKeysUnfrozen";
    EventIdEnum["UnexpectedError"] = "UnexpectedError";
    EventIdEnum["ControllerChanged"] = "ControllerChanged";
    EventIdEnum["AdminChanged"] = "AdminChanged";
    EventIdEnum["TimelockChanged"] = "TimelockChanged";
    EventIdEnum["Bridged"] = "Bridged";
    EventIdEnum["Frozen"] = "Frozen";
    EventIdEnum["Unfrozen"] = "Unfrozen";
    EventIdEnum["FrozenTx"] = "FrozenTx";
    EventIdEnum["UnfrozenTx"] = "UnfrozenTx";
    EventIdEnum["ExemptedUpdated"] = "ExemptedUpdated";
    EventIdEnum["BridgeLimitUpdated"] = "BridgeLimitUpdated";
    EventIdEnum["TxsHandled"] = "TxsHandled";
    EventIdEnum["BridgeTxScheduled"] = "BridgeTxScheduled";
    EventIdEnum["FreezeAdminAdded"] = "FreezeAdminAdded";
    EventIdEnum["FreezeAdminRemoved"] = "FreezeAdminRemoved";
    EventIdEnum["BridgeTxScheduleFailed"] = "BridgeTxScheduleFailed";
    EventIdEnum["TxRemoved"] = "TxRemoved";
    EventIdEnum["ComplianceRequirementCreated"] = "ComplianceRequirementCreated";
    EventIdEnum["ComplianceRequirementRemoved"] = "ComplianceRequirementRemoved";
    EventIdEnum["AssetComplianceReplaced"] = "AssetComplianceReplaced";
    EventIdEnum["AssetComplianceReset"] = "AssetComplianceReset";
    EventIdEnum["AssetComplianceResumed"] = "AssetComplianceResumed";
    EventIdEnum["AssetCompliancePaused"] = "AssetCompliancePaused";
    EventIdEnum["ComplianceRequirementChanged"] = "ComplianceRequirementChanged";
    EventIdEnum["TrustedDefaultClaimIssuerAdded"] = "TrustedDefaultClaimIssuerAdded";
    EventIdEnum["TrustedDefaultClaimIssuerRemoved"] = "TrustedDefaultClaimIssuerRemoved";
    EventIdEnum["BallotCreated"] = "BallotCreated";
    EventIdEnum["VoteCast"] = "VoteCast";
    EventIdEnum["BallotCancelled"] = "BallotCancelled";
    EventIdEnum["AssetPurchased"] = "AssetPurchased";
    EventIdEnum["ExemptionListModified"] = "ExemptionListModified";
    EventIdEnum["VenueCreated"] = "VenueCreated";
    EventIdEnum["VenueUpdated"] = "VenueUpdated";
    EventIdEnum["InstructionCreated"] = "InstructionCreated";
    EventIdEnum["InstructionAuthorized"] = "InstructionAuthorized";
    EventIdEnum["InstructionUnauthorized"] = "InstructionUnauthorized";
    EventIdEnum["InstructionAffirmed"] = "InstructionAffirmed";
    EventIdEnum["AffirmationWithdrawn"] = "AffirmationWithdrawn";
    EventIdEnum["InstructionRejected"] = "InstructionRejected";
    EventIdEnum["ReceiptClaimed"] = "ReceiptClaimed";
    EventIdEnum["ReceiptValidityChanged"] = "ReceiptValidityChanged";
    EventIdEnum["ReceiptUnclaimed"] = "ReceiptUnclaimed";
    EventIdEnum["VenueFiltering"] = "VenueFiltering";
    EventIdEnum["VenuesAllowed"] = "VenuesAllowed";
    EventIdEnum["VenuesBlocked"] = "VenuesBlocked";
    EventIdEnum["LegFailedExecution"] = "LegFailedExecution";
    EventIdEnum["InstructionRescheduled"] = "InstructionRescheduled";
    EventIdEnum["InstructionFailed"] = "InstructionFailed";
    EventIdEnum["InstructionExecuted"] = "InstructionExecuted";
    EventIdEnum["VenueUnauthorized"] = "VenueUnauthorized";
    EventIdEnum["VenueDetailsUpdated"] = "VenueDetailsUpdated";
    EventIdEnum["VenueTypeUpdated"] = "VenueTypeUpdated";
    EventIdEnum["FundraiserCreated"] = "FundraiserCreated";
    EventIdEnum["FundsRaised"] = "FundsRaised";
    EventIdEnum["FundraiserWindowModifed"] = "FundraiserWindowModifed";
    EventIdEnum["FundraiserClosed"] = "FundraiserClosed";
    EventIdEnum["FundraiserFrozen"] = "FundraiserFrozen";
    EventIdEnum["FundraiserUnfrozen"] = "FundraiserUnfrozen";
    EventIdEnum["TransferManagerAdded"] = "TransferManagerAdded";
    EventIdEnum["TransferManagerRemoved"] = "TransferManagerRemoved";
    EventIdEnum["ExemptionsAdded"] = "ExemptionsAdded";
    EventIdEnum["ExemptionsRemoved"] = "ExemptionsRemoved";
    EventIdEnum["FeeSet"] = "FeeSet";
    EventIdEnum["CoefficientSet"] = "CoefficientSet";
    EventIdEnum["FeeCharged"] = "FeeCharged";
    EventIdEnum["BatchInterrupted"] = "BatchInterrupted";
    EventIdEnum["BatchOptimisticFailed"] = "BatchOptimisticFailed";
    EventIdEnum["BatchCompleted"] = "BatchCompleted";
    EventIdEnum["PortfolioCreated"] = "PortfolioCreated";
    EventIdEnum["PortfolioDeleted"] = "PortfolioDeleted";
    EventIdEnum["MovedBetweenPortfolios"] = "MovedBetweenPortfolios";
    EventIdEnum["PortfolioRenamed"] = "PortfolioRenamed";
    EventIdEnum["UserPortfolios"] = "UserPortfolios";
    EventIdEnum["PortfolioCustodianChanged"] = "PortfolioCustodianChanged";
    EventIdEnum["RangeProofAdded"] = "RangeProofAdded";
    EventIdEnum["RangeProofVerified"] = "RangeProofVerified";
    EventIdEnum["Scheduled"] = "Scheduled";
    EventIdEnum["Canceled"] = "Canceled";
    EventIdEnum["Dispatched"] = "Dispatched";
    EventIdEnum["MaxDetailsLengthChanged"] = "MaxDetailsLengthChanged";
    EventIdEnum["DefaultTargetIdentitiesChanged"] = "DefaultTargetIdentitiesChanged";
    EventIdEnum["DefaultWithholdingTaxChanged"] = "DefaultWithholdingTaxChanged";
    EventIdEnum["DidWithholdingTaxChanged"] = "DidWithholdingTaxChanged";
    EventIdEnum["CaaTransferred"] = "CAATransferred";
    EventIdEnum["CaInitiated"] = "CAInitiated";
    EventIdEnum["CaLinkedToDoc"] = "CALinkedToDoc";
    EventIdEnum["CaRemoved"] = "CARemoved";
    EventIdEnum["RecordDateChanged"] = "RecordDateChanged";
    EventIdEnum["Created"] = "Created";
    EventIdEnum["RangeChanged"] = "RangeChanged";
    EventIdEnum["MetaChanged"] = "MetaChanged";
    EventIdEnum["RcvChanged"] = "RCVChanged";
    EventIdEnum["Removed"] = "Removed";
    EventIdEnum["BenefitClaimed"] = "BenefitClaimed";
    EventIdEnum["Reclaimed"] = "Reclaimed";
    EventIdEnum["MaximumSchedulesComplexityChanged"] = "MaximumSchedulesComplexityChanged";
    EventIdEnum["ScheduleCreated"] = "ScheduleCreated";
    EventIdEnum["ScheduleRemoved"] = "ScheduleRemoved";
    EventIdEnum["GroupCreated"] = "GroupCreated";
    EventIdEnum["GroupPermissionsUpdated"] = "GroupPermissionsUpdated";
    EventIdEnum["AgentAdded"] = "AgentAdded";
    EventIdEnum["AgentRemoved"] = "AgentRemoved";
    EventIdEnum["GroupChanged"] = "GroupChanged";
    EventIdEnum["AuthorizedPayingKey"] = "AuthorizedPayingKey";
    EventIdEnum["AcceptedPayingKey"] = "AcceptedPayingKey";
    EventIdEnum["RemovedPayingKey"] = "RemovedPayingKey";
    EventIdEnum["UpdatedPolyxLimit"] = "UpdatedPolyxLimit";
    EventIdEnum["ItnRewardClaimed"] = "ItnRewardClaimed";
    EventIdEnum["CustodyTransfer"] = "CustodyTransfer";
    EventIdEnum["CustodyAllowanceChanged"] = "CustodyAllowanceChanged";
    EventIdEnum["TreasuryDidSet"] = "TreasuryDidSet";
    EventIdEnum["SigningKeysAdded"] = "SigningKeysAdded";
    EventIdEnum["SigningKeysRemoved"] = "SigningKeysRemoved";
    EventIdEnum["SigningPermissionsUpdated"] = "SigningPermissionsUpdated";
    EventIdEnum["MasterKeyUpdated"] = "MasterKeyUpdated";
    EventIdEnum["CddRequirementForMasterKeyUpdated"] = "CddRequirementForMasterKeyUpdated";
    EventIdEnum["SigningKeysFrozen"] = "SigningKeysFrozen";
    EventIdEnum["SigningKeysUnfrozen"] = "SigningKeysUnfrozen";
    EventIdEnum["NewAssetRuleCreated"] = "NewAssetRuleCreated";
    EventIdEnum["AssetRuleRemoved"] = "AssetRuleRemoved";
    EventIdEnum["AssetRulesReplaced"] = "AssetRulesReplaced";
    EventIdEnum["AssetRulesReset"] = "AssetRulesReset";
    EventIdEnum["AssetRulesResumed"] = "AssetRulesResumed";
    EventIdEnum["AssetRulesPaused"] = "AssetRulesPaused";
    EventIdEnum["AssetRuleChanged"] = "AssetRuleChanged";
})(EventIdEnum = exports.EventIdEnum || (exports.EventIdEnum = {}));
var ItnRewardActionType;
(function (ItnRewardActionType) {
    ItnRewardActionType["Staking"] = "Staking";
    ItnRewardActionType["Onboarding"] = "Onboarding";
    ItnRewardActionType["PolyxTransfer"] = "PolyxTransfer";
    ItnRewardActionType["SecondaryKey"] = "SecondaryKey";
    ItnRewardActionType["ReserveTokenTicker"] = "ReserveTokenTicker";
    ItnRewardActionType["SecurityToken"] = "SecurityToken";
    ItnRewardActionType["ComplianceRequirement"] = "ComplianceRequirement";
    ItnRewardActionType["TrustedDefaultClaimIssuerAdded"] = "TrustedDefaultClaimIssuerAdded";
    ItnRewardActionType["ClaimAdded"] = "ClaimAdded";
    ItnRewardActionType["DistributeAnAsset"] = "DistributeAnAsset";
    ItnRewardActionType["AcceptATransferFromAnotherUser"] = "AcceptATransferFromAnotherUser";
    ItnRewardActionType["ConfigureAdvancedTokenRules"] = "ConfigureAdvancedTokenRules";
    ItnRewardActionType["CreateSto"] = "CreateSTO";
    ItnRewardActionType["FindInvestors"] = "FindInvestors";
    ItnRewardActionType["PortfolioCreated"] = "PortfolioCreated";
    ItnRewardActionType["AddAssetToAPortfolio"] = "AddAssetToAPortfolio";
    ItnRewardActionType["AddAPortfolioManager"] = "AddAPortfolioManager";
    ItnRewardActionType["TransferAssetFromPortfolio"] = "TransferAssetFromPortfolio";
    ItnRewardActionType["StopStakingAPortion"] = "StopStakingAPortion";
    ItnRewardActionType["StopStakingAnOperator"] = "StopStakingAnOperator";
    ItnRewardActionType["StartStakingANewOperator"] = "StartStakingANewOperator";
    ItnRewardActionType["CreateACorporateAction"] = "CreateACorporateAction";
    ItnRewardActionType["DistributeADividend"] = "DistributeADividend";
    ItnRewardActionType["AcceptADividend"] = "AcceptADividend";
    ItnRewardActionType["CreatePip"] = "CreatePip";
    ItnRewardActionType["VoteOnPip"] = "VoteOnPip";
})(ItnRewardActionType = exports.ItnRewardActionType || (exports.ItnRewardActionType = {}));
var ModuleIdEnum;
(function (ModuleIdEnum) {
    ModuleIdEnum["System"] = "system";
    ModuleIdEnum["Babe"] = "babe";
    ModuleIdEnum["Timestamp"] = "timestamp";
    ModuleIdEnum["Indices"] = "indices";
    ModuleIdEnum["Balances"] = "balances";
    ModuleIdEnum["Transactionpayment"] = "transactionpayment";
    ModuleIdEnum["Authorship"] = "authorship";
    ModuleIdEnum["Staking"] = "staking";
    ModuleIdEnum["Offences"] = "offences";
    ModuleIdEnum["Session"] = "session";
    ModuleIdEnum["Finalitytracker"] = "finalitytracker";
    ModuleIdEnum["Grandpa"] = "grandpa";
    ModuleIdEnum["Imonline"] = "imonline";
    ModuleIdEnum["Authoritydiscovery"] = "authoritydiscovery";
    ModuleIdEnum["Randomnesscollectiveflip"] = "randomnesscollectiveflip";
    ModuleIdEnum["Historical"] = "historical";
    ModuleIdEnum["Sudo"] = "sudo";
    ModuleIdEnum["Multisig"] = "multisig";
    ModuleIdEnum["Basecontracts"] = "basecontracts";
    ModuleIdEnum["Contracts"] = "contracts";
    ModuleIdEnum["Treasury"] = "treasury";
    ModuleIdEnum["Polymeshcommittee"] = "polymeshcommittee";
    ModuleIdEnum["Committeemembership"] = "committeemembership";
    ModuleIdEnum["Pips"] = "pips";
    ModuleIdEnum["Technicalcommittee"] = "technicalcommittee";
    ModuleIdEnum["Technicalcommitteemembership"] = "technicalcommitteemembership";
    ModuleIdEnum["Upgradecommittee"] = "upgradecommittee";
    ModuleIdEnum["Upgradecommitteemembership"] = "upgradecommitteemembership";
    ModuleIdEnum["Asset"] = "asset";
    ModuleIdEnum["Dividend"] = "dividend";
    ModuleIdEnum["Identity"] = "identity";
    ModuleIdEnum["Bridge"] = "bridge";
    ModuleIdEnum["Compliancemanager"] = "compliancemanager";
    ModuleIdEnum["Externalagents"] = "externalagents";
    ModuleIdEnum["Voting"] = "voting";
    ModuleIdEnum["Stocapped"] = "stocapped";
    ModuleIdEnum["Exemption"] = "exemption";
    ModuleIdEnum["Settlement"] = "settlement";
    ModuleIdEnum["Sto"] = "sto";
    ModuleIdEnum["Cddserviceproviders"] = "cddserviceproviders";
    ModuleIdEnum["Statistics"] = "statistics";
    ModuleIdEnum["Protocolfee"] = "protocolfee";
    ModuleIdEnum["Utility"] = "utility";
    ModuleIdEnum["Portfolio"] = "portfolio";
    ModuleIdEnum["Confidential"] = "confidential";
    ModuleIdEnum["Permissions"] = "permissions";
    ModuleIdEnum["Scheduler"] = "scheduler";
    ModuleIdEnum["Corporateaction"] = "corporateaction";
    ModuleIdEnum["Corporateballot"] = "corporateballot";
    ModuleIdEnum["Capitaldistribution"] = "capitaldistribution";
    ModuleIdEnum["Checkpoint"] = "checkpoint";
    ModuleIdEnum["Testnet"] = "testnet";
    ModuleIdEnum["Rewards"] = "rewards";
    ModuleIdEnum["Relayer"] = "relayer";
})(ModuleIdEnum = exports.ModuleIdEnum || (exports.ModuleIdEnum = {}));
var Order;
(function (Order) {
    Order["Asc"] = "ASC";
    Order["Desc"] = "DESC";
})(Order = exports.Order || (exports.Order = {}));
var ProposalOrderFields;
(function (ProposalOrderFields) {
    ProposalOrderFields["ProposalId"] = "proposalId";
    ProposalOrderFields["VotesCount"] = "votesCount";
})(ProposalOrderFields = exports.ProposalOrderFields || (exports.ProposalOrderFields = {}));
var ProposalStateEnum;
(function (ProposalStateEnum) {
    ProposalStateEnum["All"] = "All";
    ProposalStateEnum["Pending"] = "Pending";
    ProposalStateEnum["Rejected"] = "Rejected";
    ProposalStateEnum["Scheduled"] = "Scheduled";
    ProposalStateEnum["Failed"] = "Failed";
    ProposalStateEnum["Executed"] = "Executed";
    ProposalStateEnum["Expired"] = "Expired";
})(ProposalStateEnum = exports.ProposalStateEnum || (exports.ProposalStateEnum = {}));
var ProposalVotesOrderFields;
(function (ProposalVotesOrderFields) {
    ProposalVotesOrderFields["BlockId"] = "block_id";
    ProposalVotesOrderFields["Vote"] = "vote";
    ProposalVotesOrderFields["Weight"] = "weight";
})(ProposalVotesOrderFields = exports.ProposalVotesOrderFields || (exports.ProposalVotesOrderFields = {}));
var SettlementDirectionEnum;
(function (SettlementDirectionEnum) {
    SettlementDirectionEnum["None"] = "None";
    SettlementDirectionEnum["Incoming"] = "Incoming";
    SettlementDirectionEnum["Outgoing"] = "Outgoing";
})(SettlementDirectionEnum = exports.SettlementDirectionEnum || (exports.SettlementDirectionEnum = {}));
var SettlementResultEnum;
(function (SettlementResultEnum) {
    SettlementResultEnum["None"] = "None";
    SettlementResultEnum["Executed"] = "Executed";
    SettlementResultEnum["Failed"] = "Failed";
    SettlementResultEnum["Rejected"] = "Rejected";
})(SettlementResultEnum = exports.SettlementResultEnum || (exports.SettlementResultEnum = {}));
var SnapshotEnum;
(function (SnapshotEnum) {
    SnapshotEnum["All"] = "All";
    SnapshotEnum["InSnapshot"] = "InSnapshot";
    SnapshotEnum["NotInSnapshot"] = "NotInSnapshot";
})(SnapshotEnum = exports.SnapshotEnum || (exports.SnapshotEnum = {}));
var StakingEventIdEnum;
(function (StakingEventIdEnum) {
    StakingEventIdEnum["Bonded"] = "Bonded";
    StakingEventIdEnum["Unbonded"] = "Unbonded";
    StakingEventIdEnum["Rebond"] = "Rebond";
    StakingEventIdEnum["WithdrawUnbonded"] = "WithdrawUnbonded";
    StakingEventIdEnum["Nominated"] = "Nominated";
    StakingEventIdEnum["Reward"] = "Reward";
    StakingEventIdEnum["Slash"] = "Slash";
})(StakingEventIdEnum = exports.StakingEventIdEnum || (exports.StakingEventIdEnum = {}));
var TransactionOrderFields;
(function (TransactionOrderFields) {
    TransactionOrderFields["BlockId"] = "block_id";
    TransactionOrderFields["Address"] = "address";
    TransactionOrderFields["ModuleId"] = "module_id";
    TransactionOrderFields["CallId"] = "call_id";
})(TransactionOrderFields = exports.TransactionOrderFields || (exports.TransactionOrderFields = {}));
