// @polymathnetwork/polymesh-sdk/generated/types.js
"use strict";
/* istanbul ignore file */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CountryCode = void 0;
var CountryCode;
(function (CountryCode) {
    CountryCode["Af"] = "Af";
    CountryCode["Ax"] = "Ax";
    CountryCode["Al"] = "Al";
    CountryCode["Dz"] = "Dz";
    CountryCode["As"] = "As";
    CountryCode["Ad"] = "Ad";
    CountryCode["Ao"] = "Ao";
    CountryCode["Ai"] = "Ai";
    CountryCode["Aq"] = "Aq";
    CountryCode["Ag"] = "Ag";
    CountryCode["Ar"] = "Ar";
    CountryCode["Am"] = "Am";
    CountryCode["Aw"] = "Aw";
    CountryCode["Au"] = "Au";
    CountryCode["At"] = "At";
    CountryCode["Az"] = "Az";
    CountryCode["Bs"] = "Bs";
    CountryCode["Bh"] = "Bh";
    CountryCode["Bd"] = "Bd";
    CountryCode["Bb"] = "Bb";
    CountryCode["By"] = "By";
    CountryCode["Be"] = "Be";
    CountryCode["Bz"] = "Bz";
    CountryCode["Bj"] = "Bj";
    CountryCode["Bm"] = "Bm";
    CountryCode["Bt"] = "Bt";
    CountryCode["Bo"] = "Bo";
    CountryCode["Ba"] = "Ba";
    CountryCode["Bw"] = "Bw";
    CountryCode["Bv"] = "Bv";
    CountryCode["Br"] = "Br";
    CountryCode["Vg"] = "Vg";
    CountryCode["Io"] = "Io";
    CountryCode["Bn"] = "Bn";
    CountryCode["Bg"] = "Bg";
    CountryCode["Bf"] = "Bf";
    CountryCode["Bi"] = "Bi";
    CountryCode["Kh"] = "Kh";
    CountryCode["Cm"] = "Cm";
    CountryCode["Ca"] = "Ca";
    CountryCode["Cv"] = "Cv";
    CountryCode["Ky"] = "Ky";
    CountryCode["Cf"] = "Cf";
    CountryCode["Td"] = "Td";
    CountryCode["Cl"] = "Cl";
    CountryCode["Cn"] = "Cn";
    CountryCode["Hk"] = "Hk";
    CountryCode["Mo"] = "Mo";
    CountryCode["Cx"] = "Cx";
    CountryCode["Cc"] = "Cc";
    CountryCode["Co"] = "Co";
    CountryCode["Km"] = "Km";
    CountryCode["Cg"] = "Cg";
    CountryCode["Cd"] = "Cd";
    CountryCode["Ck"] = "Ck";
    CountryCode["Cr"] = "Cr";
    CountryCode["Ci"] = "Ci";
    CountryCode["Hr"] = "Hr";
    CountryCode["Cu"] = "Cu";
    CountryCode["Cy"] = "Cy";
    CountryCode["Cz"] = "Cz";
    CountryCode["Dk"] = "Dk";
    CountryCode["Dj"] = "Dj";
    CountryCode["Dm"] = "Dm";
    CountryCode["Do"] = "Do";
    CountryCode["Ec"] = "Ec";
    CountryCode["Eg"] = "Eg";
    CountryCode["Sv"] = "Sv";
    CountryCode["Gq"] = "Gq";
    CountryCode["Er"] = "Er";
    CountryCode["Ee"] = "Ee";
    CountryCode["Et"] = "Et";
    CountryCode["Fk"] = "Fk";
    CountryCode["Fo"] = "Fo";
    CountryCode["Fj"] = "Fj";
    CountryCode["Fi"] = "Fi";
    CountryCode["Fr"] = "Fr";
    CountryCode["Gf"] = "Gf";
    CountryCode["Pf"] = "Pf";
    CountryCode["Tf"] = "Tf";
    CountryCode["Ga"] = "Ga";
    CountryCode["Gm"] = "Gm";
    CountryCode["Ge"] = "Ge";
    CountryCode["De"] = "De";
    CountryCode["Gh"] = "Gh";
    CountryCode["Gi"] = "Gi";
    CountryCode["Gr"] = "Gr";
    CountryCode["Gl"] = "Gl";
    CountryCode["Gd"] = "Gd";
    CountryCode["Gp"] = "Gp";
    CountryCode["Gu"] = "Gu";
    CountryCode["Gt"] = "Gt";
    CountryCode["Gg"] = "Gg";
    CountryCode["Gn"] = "Gn";
    CountryCode["Gw"] = "Gw";
    CountryCode["Gy"] = "Gy";
    CountryCode["Ht"] = "Ht";
    CountryCode["Hm"] = "Hm";
    CountryCode["Va"] = "Va";
    CountryCode["Hn"] = "Hn";
    CountryCode["Hu"] = "Hu";
    CountryCode["Is"] = "Is";
    CountryCode["In"] = "In";
    CountryCode["Id"] = "Id";
    CountryCode["Ir"] = "Ir";
    CountryCode["Iq"] = "Iq";
    CountryCode["Ie"] = "Ie";
    CountryCode["Im"] = "Im";
    CountryCode["Il"] = "Il";
    CountryCode["It"] = "It";
    CountryCode["Jm"] = "Jm";
    CountryCode["Jp"] = "Jp";
    CountryCode["Je"] = "Je";
    CountryCode["Jo"] = "Jo";
    CountryCode["Kz"] = "Kz";
    CountryCode["Ke"] = "Ke";
    CountryCode["Ki"] = "Ki";
    CountryCode["Kp"] = "Kp";
    CountryCode["Kr"] = "Kr";
    CountryCode["Kw"] = "Kw";
    CountryCode["Kg"] = "Kg";
    CountryCode["La"] = "La";
    CountryCode["Lv"] = "Lv";
    CountryCode["Lb"] = "Lb";
    CountryCode["Ls"] = "Ls";
    CountryCode["Lr"] = "Lr";
    CountryCode["Ly"] = "Ly";
    CountryCode["Li"] = "Li";
    CountryCode["Lt"] = "Lt";
    CountryCode["Lu"] = "Lu";
    CountryCode["Mk"] = "Mk";
    CountryCode["Mg"] = "Mg";
    CountryCode["Mw"] = "Mw";
    CountryCode["My"] = "My";
    CountryCode["Mv"] = "Mv";
    CountryCode["Ml"] = "Ml";
    CountryCode["Mt"] = "Mt";
    CountryCode["Mh"] = "Mh";
    CountryCode["Mq"] = "Mq";
    CountryCode["Mr"] = "Mr";
    CountryCode["Mu"] = "Mu";
    CountryCode["Yt"] = "Yt";
    CountryCode["Mx"] = "Mx";
    CountryCode["Fm"] = "Fm";
    CountryCode["Md"] = "Md";
    CountryCode["Mc"] = "Mc";
    CountryCode["Mn"] = "Mn";
    CountryCode["Me"] = "Me";
    CountryCode["Ms"] = "Ms";
    CountryCode["Ma"] = "Ma";
    CountryCode["Mz"] = "Mz";
    CountryCode["Mm"] = "Mm";
    CountryCode["Na"] = "Na";
    CountryCode["Nr"] = "Nr";
    CountryCode["Np"] = "Np";
    CountryCode["Nl"] = "Nl";
    CountryCode["An"] = "An";
    CountryCode["Nc"] = "Nc";
    CountryCode["Nz"] = "Nz";
    CountryCode["Ni"] = "Ni";
    CountryCode["Ne"] = "Ne";
    CountryCode["Ng"] = "Ng";
    CountryCode["Nu"] = "Nu";
    CountryCode["Nf"] = "Nf";
    CountryCode["Mp"] = "Mp";
    CountryCode["No"] = "No";
    CountryCode["Om"] = "Om";
    CountryCode["Pk"] = "Pk";
    CountryCode["Pw"] = "Pw";
    CountryCode["Ps"] = "Ps";
    CountryCode["Pa"] = "Pa";
    CountryCode["Pg"] = "Pg";
    CountryCode["Py"] = "Py";
    CountryCode["Pe"] = "Pe";
    CountryCode["Ph"] = "Ph";
    CountryCode["Pn"] = "Pn";
    CountryCode["Pl"] = "Pl";
    CountryCode["Pt"] = "Pt";
    CountryCode["Pr"] = "Pr";
    CountryCode["Qa"] = "Qa";
    CountryCode["Re"] = "Re";
    CountryCode["Ro"] = "Ro";
    CountryCode["Ru"] = "Ru";
    CountryCode["Rw"] = "Rw";
    CountryCode["Bl"] = "Bl";
    CountryCode["Sh"] = "Sh";
    CountryCode["Kn"] = "Kn";
    CountryCode["Lc"] = "Lc";
    CountryCode["Mf"] = "Mf";
    CountryCode["Pm"] = "Pm";
    CountryCode["Vc"] = "Vc";
    CountryCode["Ws"] = "Ws";
    CountryCode["Sm"] = "Sm";
    CountryCode["St"] = "St";
    CountryCode["Sa"] = "Sa";
    CountryCode["Sn"] = "Sn";
    CountryCode["Rs"] = "Rs";
    CountryCode["Sc"] = "Sc";
    CountryCode["Sl"] = "Sl";
    CountryCode["Sg"] = "Sg";
    CountryCode["Sk"] = "Sk";
    CountryCode["Si"] = "Si";
    CountryCode["Sb"] = "Sb";
    CountryCode["So"] = "So";
    CountryCode["Za"] = "Za";
    CountryCode["Gs"] = "Gs";
    CountryCode["Ss"] = "Ss";
    CountryCode["Es"] = "Es";
    CountryCode["Lk"] = "Lk";
    CountryCode["Sd"] = "Sd";
    CountryCode["Sr"] = "Sr";
    CountryCode["Sj"] = "Sj";
    CountryCode["Sz"] = "Sz";
    CountryCode["Se"] = "Se";
    CountryCode["Ch"] = "Ch";
    CountryCode["Sy"] = "Sy";
    CountryCode["Tw"] = "Tw";
    CountryCode["Tj"] = "Tj";
    CountryCode["Tz"] = "Tz";
    CountryCode["Th"] = "Th";
    CountryCode["Tl"] = "Tl";
    CountryCode["Tg"] = "Tg";
    CountryCode["Tk"] = "Tk";
    CountryCode["To"] = "To";
    CountryCode["Tt"] = "Tt";
    CountryCode["Tn"] = "Tn";
    CountryCode["Tr"] = "Tr";
    CountryCode["Tm"] = "Tm";
    CountryCode["Tc"] = "Tc";
    CountryCode["Tv"] = "Tv";
    CountryCode["Ug"] = "Ug";
    CountryCode["Ua"] = "Ua";
    CountryCode["Ae"] = "Ae";
    CountryCode["Gb"] = "Gb";
    CountryCode["Us"] = "Us";
    CountryCode["Um"] = "Um";
    CountryCode["Uy"] = "Uy";
    CountryCode["Uz"] = "Uz";
    CountryCode["Vu"] = "Vu";
    CountryCode["Ve"] = "Ve";
    CountryCode["Vn"] = "Vn";
    CountryCode["Vi"] = "Vi";
    CountryCode["Wf"] = "Wf";
    CountryCode["Eh"] = "Eh";
    CountryCode["Ye"] = "Ye";
    CountryCode["Zm"] = "Zm";
    CountryCode["Zw"] = "Zw";
    CountryCode["Bq"] = "Bq";
    CountryCode["Cw"] = "Cw";
    CountryCode["Sx"] = "Sx";
})(CountryCode = exports.CountryCode || (exports.CountryCode = {}));
