// @polymathnetwork/polymesh-sdk/types/internal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AllowanceOperation = exports.InstructionStatus = exports.InstructionAffirmationOperation = exports.TrustedClaimIssuerOperation = exports.ClaimOperation = void 0;
var ClaimOperation;
(function (ClaimOperation) {
    ClaimOperation["Revoke"] = "Revoke";
    ClaimOperation["Add"] = "Add";
    ClaimOperation["Edit"] = "Edit";
})(ClaimOperation = exports.ClaimOperation || (exports.ClaimOperation = {}));
var TrustedClaimIssuerOperation;
(function (TrustedClaimIssuerOperation) {
    TrustedClaimIssuerOperation["Remove"] = "Remove";
    TrustedClaimIssuerOperation["Add"] = "Add";
    TrustedClaimIssuerOperation["Set"] = "Set";
})(TrustedClaimIssuerOperation = exports.TrustedClaimIssuerOperation || (exports.TrustedClaimIssuerOperation = {}));
var InstructionAffirmationOperation;
(function (InstructionAffirmationOperation) {
    InstructionAffirmationOperation["Affirm"] = "Affirm";
    InstructionAffirmationOperation["Withdraw"] = "Withdraw";
    InstructionAffirmationOperation["Reject"] = "Reject";
})(InstructionAffirmationOperation = exports.InstructionAffirmationOperation || (exports.InstructionAffirmationOperation = {}));
var InstructionStatus;
(function (InstructionStatus) {
    InstructionStatus["Pending"] = "Pending";
    InstructionStatus["Unknown"] = "Unknown";
    InstructionStatus["Failed"] = "Failed";
})(InstructionStatus = exports.InstructionStatus || (exports.InstructionStatus = {}));
var AllowanceOperation;
(function (AllowanceOperation) {
    AllowanceOperation["Set"] = "Set";
    AllowanceOperation["Increase"] = "Increase";
    AllowanceOperation["Decrease"] = "Decrease";
})(AllowanceOperation = exports.AllowanceOperation || (exports.AllowanceOperation = {}));
