// @polymathnetwork/polymesh-sdk/api/procedures/inviteExternalAgent.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.inviteExternalAgent = exports.prepareStorage = exports.getAuthorization = exports.prepareInviteExternalAgent = void 0;
const utils_1 = require("./utils");
const internal_1 = require("../../internal");
const types_1 = require("../../types");
const conversion_1 = require("../../utils/conversion");
const internal_2 = require("../../utils/internal");
/**
 * @hidden
 */
function prepareInviteExternalAgent(args) {
    return __awaiter(this, void 0, void 0, function* () {
        const { context: { polymeshApi: { tx: { identity }, }, }, context, storage: { asset }, } = this;
        const { ticker, target, permissions, expiry = null } = args;
        const issuer = yield context.getSigningIdentity();
        const targetIdentity = yield context.getIdentity(target);
        const currentAgents = yield asset.permissions.getAgents();
        const isAgent = !!currentAgents.find(({ agent }) => agent.isEqual(targetIdentity));
        if (isAgent) {
            throw new internal_1.PolymeshError({
                code: types_1.ErrorCode.NoDataChange,
                message: 'The target Identity is already an External Agent',
            });
        }
        const rawSignatory = (0, conversion_1.signerValueToSignatory)({ type: types_1.SignerType.Identity, value: (0, conversion_1.signerToString)(target) }, context);
        let postTransactionAuthorization;
        let rawAuthorizationData;
        // helper to transform permissions into the relevant Authorization
        const createBecomeAgentData = (value) => ({
            type: types_1.AuthorizationType.BecomeAgent,
            value,
        });
        if (permissions instanceof internal_1.KnownPermissionGroup || permissions instanceof internal_1.CustomPermissionGroup) {
            postTransactionAuthorization = createBecomeAgentData(permissions);
            rawAuthorizationData = (0, conversion_1.authorizationToAuthorizationData)(postTransactionAuthorization, context);
        }
        else {
            // We know this procedure returns a PostTransactionValue, so this assertion is necessary
            const createGroupResult = (yield this.addProcedure((0, internal_1.createGroup)(), {
                ticker,
                permissions,
            }));
            postTransactionAuthorization = createGroupResult.transform(createBecomeAgentData);
            rawAuthorizationData = postTransactionAuthorization.transform(authorization => (0, conversion_1.authorizationToAuthorizationData)(authorization, context));
        }
        const rawExpiry = (0, internal_2.optionize)(conversion_1.dateToMoment)(expiry, context);
        const [auth] = this.addTransaction({
            transaction: identity.addAuthorization,
            resolvers: [
                (0, utils_1.createAuthorizationResolver)(postTransactionAuthorization, issuer, targetIdentity, expiry, context),
            ],
            args: [rawSignatory, rawAuthorizationData, rawExpiry],
        });
        return auth;
    });
}
exports.prepareInviteExternalAgent = prepareInviteExternalAgent;
/**
 * @hidden
 */
function getAuthorization() {
    const { storage: { asset }, } = this;
    return {
        permissions: {
            transactions: [types_1.TxTags.identity.AddAuthorization],
            assets: [asset],
            portfolios: [],
        },
    };
}
exports.getAuthorization = getAuthorization;
/**
 * @hidden
 */
function prepareStorage({ ticker }) {
    const { context } = this;
    return {
        asset: new internal_1.Asset({ ticker }, context),
    };
}
exports.prepareStorage = prepareStorage;
/**
 * @hidden
 */
const inviteExternalAgent = () => new internal_1.Procedure(prepareInviteExternalAgent, getAuthorization, prepareStorage);
exports.inviteExternalAgent = inviteExternalAgent;
